File: 2018/day_1/solution.clj
#!/usr/bin/env lein exec
(require '[clojure.string :as str])

(def input (map read-string (str/split-lines (slurp *in*))))
(defn freq-seq [input] (reductions + (cycle input)))
(defn reduce-until-duplicate [freq-stream] (reduce #(if (contains? %1 %2) (reduced %2) (conj %1 %2)) #{} freq-stream))

(println (reduce + input))
(println (reduce-until-duplicate (freq-seq input)))

File: 2018/day_1/solution.exs
#! /usr/bin/env elixir

input = IO.stream(:stdio, :line)
|> Enum.map(&String.trim/1)
|> Enum.map(&String.to_integer/1)

# part 1
input |> Enum.sum |> IO.inspect

# part 2

# reduce the stream of frequencies into a set of frequencies seen before
# halt on a duplicate and return it
reduce_until_duplicate = fn x, {acc_freq, acc_map} ->
    # IO.inspect x
    acc_freq = acc_freq + x
    if MapSet.member?(acc_map, acc_freq) do
        {:halt, acc_freq}
    else
        {:cont, {acc_freq, MapSet.put(acc_map, acc_freq)}}
    end
end

input|> Stream.cycle
|> Enum.reduce_while(
    {0, MapSet.new()},
    reduce_until_duplicate
    )
|> IO.inspect
